flink-kubernetes/src/main/java/org/apache/flink/kubernetes/KubernetesResourceManager.java

```diff
@@ -237,57 +230,73 @@ private void recoverWorkerNodesFromPreviousAttempts() throws ResourceManagerExce
 			++currentMaxAttemptId);
 	}
 
-	private void requestKubernetesPod() {
-		numPendingPodRequests++;
+	private void requestKubernetesPod(WorkerResourceSpec workerResourceSpec) {
+		final KubernetesTaskManagerParameters parameters =
+			createKubernetesTaskManagerParameters(workerResourceSpec);
+
+		final KubernetesPod taskManagerPod =
+			KubernetesTaskManagerFactory.createTaskManagerComponent(parameters);
+		kubeClient.createTaskManagerPod(taskManagerPod);
+
+		podWorkerResources.put(parameters.getPodName(), workerResourceSpec);
+		final int pendingWorkerNum = notifyNewWorkerRequested(workerResourceSpec);
 
 		log.info("Requesting new TaskManager pod with <{},{}>. Number pending requests {}.",
-			defaultMemoryMB,
-			defaultCpus,
-			numPendingPodRequests);
+			parameters.getTaskManagerMemoryMB(),
+			parameters.getTaskManagerCPU(),
+			pendingWorkerNum);
+		log.info("TaskManager {} will be started with {}.", parameters.getPodName(), workerResourceSpec);
+	}
+
+	private KubernetesTaskManagerParameters createKubernetesTaskManagerParameters(WorkerResourceSpec workerResourceSpec) {
+		final TaskExecutorProcessSpec taskExecutorProcessSpec =
+			TaskExecutorProcessUtils.processSpecFromWorkerResourceSpec(flinkConfig, workerResourceSpec);
 
 		final String podName = String.format(
 			TASK_MANAGER_POD_FORMAT,
 			clusterId,
 			currentMaxAttemptId,
 			++currentMaxPodId);
 
+		final ContaineredTaskManagerParameters taskManagerParameters =
+			ContaineredTaskManagerParameters.create(flinkConfig, taskExecutorProcessSpec);
+
 		final String dynamicProperties =
 			BootstrapTools.getDynamicPropertiesAsString(flinkClientConfig, flinkConfig);
 
-		final KubernetesTaskManagerParameters kubernetesTaskManagerParameters = new KubernetesTaskManagerParameters(
+		return new KubernetesTaskManagerParameters(
 			flinkConfig,
 			podName,
 			dynamicProperties,
 			taskManagerParameters);
-
-		final KubernetesPod taskManagerPod =
-			KubernetesTaskManagerFactory.createTaskManagerComponent(kubernetesTaskManagerParameters);
-
-		log.info("TaskManager {} will be started with {}.", podName, taskExecutorProcessSpec);
-		kubeClient.createTaskManagerPod(taskManagerPod);
 	}
 
 	/**
 	 * Request new pod if pending pods cannot satisfy pending slot requests.
 	 */
-	private void requestKubernetesPodIfRequired() {
-		final int requiredTaskManagers = getNumberRequiredTaskManagers();
+	private void requestKubernetesPodIfRequired(WorkerResourceSpec workerResourceSpec) {
+		final int pendingWorkerNum = getNumPendingWorkersFor(workerResourceSpec);
+		int requiredTaskManagers = getRequiredResources().get(workerResourceSpec);
 
-		while (requiredTaskManagers > numPendingPodRequests) {
-			requestKubernetesPod();
+		while (requiredTaskManagers-- > pendingWorkerNum) {
```

Comment:
If, then I would not decrease `requiredTaskManager` but increase `pendingWorkerNum` because that is what `requestKubernetesPod` does.

Ideally, the while condition should look like `while (requiredTaskManagers > getNumPendingWorkersFor(workerResourceSpec))` because now we implicitly encode that `requestKubernetesPod` will increase `getNumPendingWorkersFor` by one. I actually think that this should also not perform worse because usually this loop will only be executed once.